### PR TITLE
V3.2

### PR DESCRIPTION
**Pull request by avarabyeu**

### CODE REVIEW

File: src/main/java/com/epam/ta/reportportal/core/launch/impl/GetLaunchHandler.java [hunk 1]
```diff
@@ -129,6 +129,17 @@ public LaunchResource getLaunchByName(String project, Pageable pageable, Filter
...
 
+    @Override
+    public com.epam.ta.reportportal.ws.model.Page<LaunchResource> getLatestLaunches(String projectName, Filter filter,
+                                                                                    Pageable pageable) {
+        validateModeConditions(filter);
+        filter.addCondition(new FilterCondition(EQUALS, false, DEFAULT.toString(), Launch.MODE_CRITERIA));
+        Page<LaunchResource> resources = launchRepository.findLatestLaunches(projectName, filter, pageable)
+                .map(launchResourceAssembler::toResource);
+        return new com.epam.ta.reportportal.ws.model.Page<>(resources.getContent(), resources.getSize(),
+                resources.getNumber() + 1, resources.getTotalElements(), resources.getTotalPages());
```
Comment (pictureb00k) [inline src/main/java/com/epam/ta/reportportal/core/launch/impl/GetLaunchHandler.java]: leave comment here about +1

File: src/main/java/com/epam/ta/reportportal/core/externalsystem/handler/impl/GetTicketHandler.java [hunk 2]
```diff
@@ -73,11 +73,29 @@ public Ticket getTicket(String ticketId, String projectName, String systemId) {
...
+
+    @Override
+    public List<String> getAllowableIssueTypes(String projectName, String systemId) {
+        validateProject(projectName);
+        ExternalSystem system = validateExternalSystem(systemId);
+        ExternalSystemStrategy externalSystemStrategy = strategyProvider.getStrategy(system.getExternalSystemType().name());
+        return externalSystemStrategy.getIssueTypes(system);
+    }
+
+    private Project validateProject(String projectName) {
```
Comment (avarabyeu) [inline src/main/java/com/epam/ta/reportportal/core/externalsystem/handler/impl/GetTicketHandler.java]: exists?

File: src/main/java/com/epam/ta/reportportal/core/item/StartTestItemHandlerImpl.java [hunk 3]
```diff
@@ -40,22 +32,33 @@
...
- * 
+ *
  */
 @Service
 class StartTestItemHandlerImpl implements StartTestItemHandler {
 	private TestItemRepository testItemRepository;
 	private LaunchRepository launchRepository;
 	private Provider<TestItemBuilder> testItemBuilder;
 	private LogRepository logRepository;
+	private String projectName;
```
Comment (avarabyeu) [inline src/main/java/com/epam/ta/reportportal/core/item/StartTestItemHandlerImpl.java]: WHAT???!!! Assigning value to class-level variable in singleton bean from?

File: src/main/java/com/epam/ta/reportportal/core/item/TestItemUniqueIdGenerator.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ * on Base64 encoding and includes information about project,
+ * name of item's launch, full path of item's parent names,
+ * item name and parameters.
+ *
+ * @author Pavel_Bortnik
+ */
+@Service
+public class TestItemUniqueIdGenerator implements UniqueIdGenerator {
+
+    private static final Base64.Encoder encoder = Base64.getEncoder();
```
Comment (avarabyeu) [inline src/main/java/com/epam/ta/reportportal/core/item/TestItemUniqueIdGenerator.java]: static final in UPPER_CASE

File: src/main/java/com/epam/ta/reportportal/core/item/TestItemUniqueIdGenerator.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private String prepareForEncoding(TestItem testItem, String projectName) {
+        String launchName = launchRepository.findNameNumberAndModeById(testItem.getLaunchRef()).getName();
+        List<String> pathNames = getPathNames(testItem.getPath());
+        String itemName = testItem.getName();
+        List<Parameter> parameters = Optional.ofNullable(testItem.getParameters()).orElse(Collections.emptyList());
+        StringJoiner joiner = new StringJoiner(";");
+        joiner.add(SECRET).add(projectName).add(launchName);
+        if (CollectionUtils.isEmpty(pathNames)) {
```
Comment (avarabyeu) [inline src/main/java/com/epam/ta/reportportal/core/item/TestItemUniqueIdGenerator.java]: seems like there should be isNOTEmpty

File: src/main/java/com/epam/ta/reportportal/core/item/merge/strategy/AbstractSuiteMergeStrategy.java [hunk 6]
```diff
@@ -95,35 +97,54 @@ protected void mergeAllChildItems(TestItem testItemParent) {
...
+            target.setUniqueId(source.getUniqueId());
+        }
         testItemRepository.save(target);
     }
+
+    /**
+     * Defines start time as the earliest and the end time as latest
+     */
+    private TestItem updateTime(TestItem target, TestItem source) {
+        target.setStartTime(target.getStartTime().compareTo(source.getStartTime()) < 0 ?
```
Comment (avarabyeu) [inline src/main/java/com/epam/ta/reportportal/core/item/merge/strategy/AbstractSuiteMergeStrategy.java]: date.after() date.before()??

File: src/main/java/com/epam/ta/reportportal/core/item/merge/strategy/AbstractSuiteMergeStrategy.java [hunk 7]
```diff
@@ -95,35 +97,54 @@ protected void mergeAllChildItems(TestItem testItemParent) {
...
+                target.getStartTime() : source.getStartTime());
+        target.setEndTime(target.getEndTime().compareTo(source.getEndTime()) > 0 ?
+                target.getEndTime() : source.getEndTime());
+        return target;
+    }
+
+    private Set<String> mergeTags(@Nullable Set<String> first, @Nullable Set<String> second) {
+        return Stream.concat(Optional.ofNullable(first).orElse(Collections.emptySet()).stream(),
```
Comment (avarabyeu) [inline src/main/java/com/epam/ta/reportportal/core/item/merge/strategy/AbstractSuiteMergeStrategy.java]: i think ternary operator would seem easier here.
Stream.concat(first != null ? first : Collections.emptySet, ...)

File: src/main/java/com/epam/ta/reportportal/core/item/StartTestItemHandlerImpl.java (hunk 3 above)
Comment (pictureb00k) [inline src/main/java/com/epam/ta/reportportal/core/item/StartTestItemHandlerImpl.java]: fixed

**Pull request merged**